lib/spack/spack/cmd/uninstall.py

```diff
@@ -158,19 +159,27 @@ def installed_dependents(specs, env):
 
     env_hashes = set(env.all_hashes()) if env else set()
 
-    all_specs_in_db = spack.store.db.query()
+    # Ensure we stop traversal at input specs.
+    visited = set(s.dag_hash() for s in specs)
 
     for spec in specs:
-        installed = [x for x in all_specs_in_db if spec in x]
-
-        # separate installed dependents into dpts in this environment and
-        # dpts that are outside this environment
-        for dpt in installed:
-            if dpt not in specs:
-                if dpt.dag_hash() in env_hashes:
-                    active_dpts.setdefault(spec, set()).add(dpt)
-                else:
-                    outside_dpts.setdefault(spec, set()).add(dpt)
+        for dpt in traverse.traverse_nodes(
+            spec.dependents(deptype="all"),
+            direction="parents",
+            visited=visited,
+            deptype="all",
+            root=True,
+            key=lambda s: s.dag_hash(),
+        ):
+            hash = dpt.dag_hash()
+            # Ensure that all the specs we get are installed
+            record = spack.store.db.query_local_by_spec_hash(hash)
+            if record is None or not record.installed:
+                continue
+            if hash in env_hashes:
+                active_dpts.setdefault(spec, set()).add(dpt)
+            else:
+                outside_dpts.setdefault(spec, set()).add(dpt)
```

Comment:
If I read the code correctly, these lists are now computed differently, and won't match the docstring description (if A is listed as a dependent of B it won't be listed anymore as a dependent of anything else). This though will only matter for error messages as they are now, not for actual installations. Also, the error messages in this PR will just be less complete - not incorrect.

On the other hand we avoid doing the same traversals over and over, which is a win in terms of performance. 

If my understanding is correct this LGTM.

Reply:
Ah, I didn't check the docstring. So yes, listing specs at most once is intended behavior: it ensures linear runtime and smaller error messages, in particular when doing `spack uninstall --all`.

Reply:
For differences in the error message using the same store, with `develop`:
![Screenshot from 2022-11-21 14-56-12](https://user-images.githubusercontent.com/4199709/203073348-eb8c850c-a611-4e8b-a76b-bee3a2554bdf.png)

with this PR:

![Screenshot from 2022-11-21 14-57-17](https://user-images.githubusercontent.com/4199709/203073462-ca80ccc9-c8ef-486a-b94d-578859496cda.png)

Reply:
Actually uninstalling everything in that store from outside environments, on `develop`:
```console
$ time spack uninstall -ay
[ ... ]
real	0m6,314s
user	0m5,957s
sys	0m0,327s
```
and with this PR:
```console
$ time spack uninstall -ay
[ ... ]
real	0m3,821s
user	0m3,492s
sys	0m0,308s
```